### ISSUE TITLE
JNI ERROR (app bug): local reference table overflow (max=512)

### ISSUE THREAD

**Issue by capezzbr** [LiquidPlayer/LiquidCore#44]
- Lib version: `0.4.4`
- Doesn't happen on version: `0.2.2`
- I'm currently invoking

```java
context.evaluateScript(src, name, 0);
```

where `src` is the content of a javascript file I'm evaluating in order to boot my JS environment.

---

Exception details

```
JNI ERROR (app bug): local reference table overflow (max=512)
local reference table dump:
  Last 10 entries (of 512):
      511: 0x12c41d00 java.lang.String "org/liquidplayer... (38 chars)
      510: 0x12c9dd60 org.liquidplayer.javascript.JNIReturnObject
      509: 0x12cbca00 java.lang.Class<org.liquidplayer.javascript.JNIReturnObject>
      508: 0x12cfcbf0 java.lang.String "org/liquidplayer... (43 chars)
      507: 0x12cbdf00 java.lang.Class<org.liquidplayer.javascript.JNIObject>
      506: 0x12c41d60 java.lang.String "org/liquidplayer... (37 chars)
      505: 0x12c71a10 org.liquidplayer.javascript.JNIJSValue
      504: 0x12ca9820 java.lang.Class<org.liquidplayer.javascript.JNIJSValue>
      503: 0x12c3ce80 java.lang.String "org/liquidplayer... (38 chars)
      502: 0x12c71a20 org.liquidplayer.javascript.JNIJSValue
  Summary:
       26 of org.liquidplayer.javascript.JNIJSObject (13 unique instances)
      102 of org.liquidplayer.javascript.JNIJSValue (84 unique instances)
       36 of org.liquidplayer.javascript.JNIReturnObject (36 unique instances)
        1 of org.liquidplayer.javascript.JNIJSValue[] (2 elements)
      174 of java.lang.String (174 unique instances)
      172 of java.lang.Class (5 unique instances)
        1 of java.lang.String[] (3 elements)

Runtime aborting...
Aborting thread:
```

**Comment by ericwlange** [comment]
Hmm.  Do you have any function callbacks that might be either circular or very deep?  Example:

Java:
```java
final JSContext ctx = new JSContext();
JSFunction javafunc = new JSFunction(ctx, "javafunc") {
    public void javafunc() {
        ctx.evaluateScript("jsfunc()");
    }
}
ctx.property("javafunc", javafunc);
ctx.evaluateScript(js_code_as_below);
```

Javascript:
```javascript
function jsfunc() {
   javafunc();
}
```

Basically anything that calls back and forth between Java and JavaScript triggered from a single call can cause this if the call stack is too deep.  Do you have a simple failing test?

**Comment by ericwlange** [comment]
Incidentally, changing `jsfunc` as follows:

```javascript
async function call_async(func) {
   await new Promise(resolve=>setTimeout(()=>{func(); resolve();},0);
}

function jsfunc() {
    call_async(javafunc);
}
```
makes this particular issue go away, although, of course, the infinite circular callbacks will still occur.  It just won't overflow the call stack.   But if you are making very deep calls, having some of them call asynchronously can fix this.

**Comment by capezzbr** [comment]
@ericwlange thanks for your reply! I thought about that too and tried to avoid recursive functions, but the interesting fact is that the same JS/Java code works smoothly on `0.2.2` so I was wondering:
- Was this changed on purpose to improve performances/memory consumption?
- Can't we extend the stack to be a bit bigger? Or would that ruin performances?

**Comment by ericwlange** [comment]
@capezzbr, @buddhirajapraksh has a patch for this in issue #45.  I haven't evaluated it yet, but it looks promising at first pass.  Maybe see if this helps.  I will evaluate the patch tonight and push a new release if all looks good.

**Comment by ericwlange** [comment]
Yes, it looks like @buddhirajapraksh has located the bug.  I created a weak reference but never delete that reference.  Will patch it.

**Comment by capezzbr** [comment]
@ericwlange thanks for getting back to me and @buddhirajapraksh awesome work on the path! Thanks a lot guys, I'll definitely checkout the new release once is out to confirm if it solves the issue 🚀

**Comment by buddhirajapraksh** [comment]
@ericwlange, @capezzbr, the patch may not solve your issue as it pertains to weak global refs while exception you got suggests it's a local ref issue. Per JNI documentation, local refs are deleted when stack unwinds, so a deep recursion as suggested by @ericwlange may cause issues. Therefore deep recursions crossing JS and Java boundaries should be avoided.

Please take this information with pinch of salt, as I do not fully understand the JNI world issues. There is no harm in trying the patch though.

**Comment by capezzbr** [comment]
I've created a test project to reproduce the issue, it seems there is a problem when invoking a native function from JS more than 512 times.  Below is the code to reproduce the issue:

```kotlin
val context = JSContext()
context.property("nativeLog", object : JSFunction(context, "nativeLog") {
    fun nativeLog(message: String) {
        Log.i("", message);
    }
})

val script = "for (var i=0; i < 600; i++) { nativeLog('Call number ' + i); }"

context.evaluateScript(script, "script.js", 0)
```

Running the same piece of code on version `0.2.2` works fine.

**Comment by ericwlange** [comment]
@capezzbr Thanks.  I can verify the failure with this test.

@buddhirajapraksh you are correct this does not solve the problem, but was in fact a different bug that you located.

Digging in.  Stay tuned

**Comment by buddhirajapraksh** [comment]
With attached patch I am able to run the loop till 255 (earlier it was barfing at < 70). Major issue left now is how to deal with JNIReturnObject and JNIJSValue

```
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132] JNI ERROR (app bug): local reference table overflow (max=512)
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132] local reference table dump:
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]   Last 10 entries (of 511):
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       510: 0x12cd7a00 java.lang.Class<org.liquidplayer.javascript.JNIReturnObject>
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       509: 0x12ddb2d0 org.liquidplayer.javascript.JNIJSValue
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       508: 0x12dcf8e0 org.liquidplayer.javascript.JNIReturnObject
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       507: 0x12ddb260 org.liquidplayer.javascript.JNIJSValue
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       506: 0x12dcf860 org.liquidplayer.javascript.JNIReturnObject
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       505: 0x12ddb1f0 org.liquidplayer.javascript.JNIJSValue
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       504: 0x12dcf7e0 org.liquidplayer.javascript.JNIReturnObject
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       503: 0x12ddb180 org.liquidplayer.javascript.JNIJSValue
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       502: 0x12dcf760 org.liquidplayer.javascript.JNIReturnObject
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       501: 0x12ddb110 org.liquidplayer.javascript.JNIJSValue
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]   Summary:
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       254 of org.liquidplayer.javascript.JNIJSValue (254 unique instances)
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]         1 of android.app.Instrumentation$InstrumentationThread
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]       254 of org.liquidplayer.javascript.JNIReturnObject (254 unique instances)
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]         1 of org.liquidplayer.javascript.JNIJSValue[] (1 elements)
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132]         1 of java.lang.Class
02-03 01:49:06.272 13318 13333 F art     : art/runtime/indirect_reference_table.cc:132] 
```
[localrefs.patch.log](https://github.com/LiquidPlayer/LiquidCore/files/1691611/localrefs.patch.log)

**Comment by ericwlange** [comment]
I got to 505 with the following patch:
[JNI.patch.txt](https://github.com/LiquidPlayer/LiquidCore/files/1691676/JNI.patch.txt)

Only thing left is `JNIJSValue`

```
2-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115] JNI ERROR (app bug): local reference table overflow (max=512)
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115] local reference table dump:
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]   Last 10 entries (of 511):
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       510: 0x12d2e900 java.lang.Class<org.liquidplayer.javascript.JNIReturnObject>
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       509: 0x12d91c00 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       508: 0x12d91b70 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       507: 0x12d91ae0 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       506: 0x12d91a50 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       505: 0x12d919c0 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       504: 0x12d91930 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       503: 0x12d918a0 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       502: 0x12d91810 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       501: 0x12d91780 org.liquidplayer.javascript.JNIJSValue
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]   Summary:
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]         1 of android.app.Instrumentation$InstrumentationThread
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]       507 of org.liquidplayer.javascript.JNIJSValue (507 unique instances)
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]         1 of org.liquidplayer.javascript.JNIReturnObject
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]         1 of org.liquidplayer.javascript.JNIJSValue[] (1 elements)
02-03 11:07:27.467 21362-21375/org.liquidplayer.node.test A/art: art/runtime/indirect_reference_table.cc:115]         1 of java.lang.Class
0
```

**Comment by ericwlange** [comment]
[This patch](https://github.com/LiquidPlayer/LiquidCore/files/1691787/JNI2.patch.txt) solves the problem.  I updated the test to include a return value, which also caused problems.  I was able to bump it to 6000 iterations and it still works.  I will publish an 0.4.5 update.  Thanks all for the help!

**Comment by ericwlange** [comment]
This is now fixed in 0.4.5

**Comment by capezzbr** [comment]
Hey @ericwlange I can confirm that solved the issue, awesome job 👏
